GVFS/GVFS.Common/GVFSEnlistmentHealthCalculator.cs

```diff
@@ -0,0 +1,258 @@
+                        {
+                            this.IncreaseDictionaryCounterByKey(directoryTracking, topDir);
+                        }
+                    }
+                }
+            }
+
+            return count;
+        }
+
+        /// <summary>
+        /// Trim the relative path to a directory from the front of a specified path which is its child
+        /// </summary>
+        /// <remarks>Precondition: Only even run in the context of TrimDirectoryFromPath(child, parentDirectory)</remarks>
+        /// <param name="path">The path being trimmed</param>
+        /// <param name="directoryTarget">The directory target whose path to trim from the path</param>
+        /// <returns>The newly formatted path with the directory trimmed</returns>
+        private string TrimDirectoryFromPath(string path, string directoryTarget)
+        {
+            return path.Substring(directoryTarget.Length);
+        }
+
+        private void IncreaseDictionaryCounterByKey(Dictionary<string, int> countingDictionary, string key)
+        {
+            if (!countingDictionary.TryGetValue(key, out int count))
+            {
+                count = 0;
+            }
+
+            countingDictionary[key] = ++count;
+        }
+
+        private decimal CalculateHealthMetric(int hydratedFileCount, int totalFileCount)
+        {
+            if (totalFileCount == 0)
+            {
+                return 0;
+            }
+
+            return (decimal)hydratedFileCount / (decimal)totalFileCount;
+        }
+
+        public class GVFSEnlistmentPathData
+        {
+            public List<string> GitFolderPaths;
+            public List<string> GitFilePaths;
+            public List<string> PlaceholderFolderPaths;
+            public List<string> PlaceholderFilePaths;
+            public List<string> ModifiedFolderPaths;
+            public List<string> ModifiedFilePaths;
+
+            public GVFSEnlistmentPathData()
+            {
+                this.GitFolderPaths = new List<string>();
+                this.GitFilePaths = new List<string>();
+                this.PlaceholderFolderPaths = new List<string>();
+                this.PlaceholderFilePaths = new List<string>();
+                this.ModifiedFolderPaths = new List<string>();
+                this.ModifiedFilePaths = new List<string>();
+            }
+
+            public void AddSkipWorkTreeFilePaths(List<string> skipWorkTreeFilePaths)
+            {
+                this.ModifiedFilePaths.Union(skipWorkTreeFilePaths);
+            }
+
+            public void NormalizeAllPaths()
+            {
+                this.NormalizePaths(this.GitFolderPaths);
+                this.NormalizePaths(this.GitFilePaths);
+                this.NormalizePaths(this.PlaceholderFolderPaths);
+                this.NormalizePaths(this.PlaceholderFilePaths);
+                this.NormalizePaths(this.ModifiedFolderPaths);
+                this.NormalizePaths(this.ModifiedFilePaths);

```

Comment:
Potential performance improvement (no need to block the PR on this comment):

Are any of these groups of paths are already guaranteed to be in the format we want (so that we wouldn't need to normalize them)?  If so, we could avoid having to normalize all of the paths,

Reply:
I'll investigate this later, it's definitely possible!